hdfs/src/object_store/hdfs.rs

```diff
@@ -440,6 +429,48 @@ fn convert_walkdir_result(
     }
 }
 
+/// Range requests with a gap less than or equal to this,
+/// will be coalesced into a single request by [`coalesce_ranges`]
+pub const HDFS_COALESCE_DEFAULT: usize = 1024 * 1024;
+
+/// Takes a function to fetch ranges and coalesces adjacent ranges if they are
+/// less than `coalesce` bytes apart. Out of order `ranges` are not coalesced
+pub async fn coalesce_ranges<F, Fut>(
+    ranges: &[std::ops::Range<usize>],
+    mut fetch: F,
+    coalesce: usize,
+) -> Result<Vec<Bytes>>
+where
+    F: FnMut(std::ops::Range<usize>) -> Fut,
+    Fut: std::future::Future<Output = Result<Bytes>>,
+{
+    let mut ret = Vec::with_capacity(ranges.len());
+    let mut start_idx = 0;
+    let mut end_idx = 1;
+
+    while start_idx != ranges.len() {
```

Comment:
We may need to sort the ranges before checking.

Reply:
👍 fix in [add sort](https://github.com/datafusion-contrib/datafusion-objectstore-hdfs/pull/10/commits/8fd32b61e9e2cfa63ca81a4482fab95ea66b9895)

Reply:
@yahoNanJing